pallets/dactr/src/extensions/check_app_id.rs

```diff
@@ -27,49 +28,82 @@ use crate::{Call as DACall, Config, Pallet};
 ///
 #[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
 #[scale_info(skip_type_params(T))]
-pub struct CheckAppId<T: Config + Send + Sync>(pub AppId, sp_std::marker::PhantomData<T>);
+pub struct CheckAppId<T: DAConfig + UtilityConfig + Send + Sync>(
+	pub AppId,
+	sp_std::marker::PhantomData<T>,
+);
 
 impl<T> CheckAppId<T>
 where
-	T: Config + Send + Sync,
-	T::RuntimeCall: IsSubType<DACall<T>>,
+	T: DAConfig + UtilityConfig + Send + Sync,
+	<T as frame_system::Config>::RuntimeCall:
+		IsSubType<DACall<T>> + IsSubType<UtilityCall<T>> + From<UtilityCall<T>>,
 {
 	/// utility constructor. Used only in client/factory code.
 	pub fn from(app_id: AppId) -> Self { Self(app_id, sp_std::marker::PhantomData) }
 
+	/* 	pub fn validate_batch_call(&self, call: &<T as pallet_utility::Config>::RuntimeCall) {} */
+
 	/// Transaction validation:
-	///  - Only `DataAvailability::submit_data(..)` extrinsic can use `AppId != 0`. Any other call
-	///  must use `AppId == 0`.
-	///  - It validates that `AppId` is already registered.
-	pub fn do_validate(&self, call: &T::RuntimeCall) -> TransactionValidity {
-		match call.is_sub_type() {
-			// Only `dactrl::submit_data` can use `AppId != 0`.
-			Some(DACall::<T>::submit_data { .. }) => {
-				let next_app_id = <Pallet<T>>::peek_next_application_id();
-				ensure!(
-					self.app_id() < next_app_id,
-					InvalidTransaction::Custom(InvalidTransactionCustomId::InvalidAppId as u8)
-				);
-			},
-			_ => {
-				// Any other call must use `AppId == 0`.
-				ensure!(
-					self.app_id().0 == 0,
-					InvalidTransaction::Custom(InvalidTransactionCustomId::ForbiddenAppId as u8)
-				);
+	///  - `DataAvailability::submit_data(..)` extrinsic can use `AppId != 0`.
+	///  - `Utility::batch/batch_all/force_batch(..)` extrinsic can use `AppId != 0` If the wrapped calls are ALL `DataAvailability::submit_data(..)`.
+	///  - Any other call must use `AppId == 0`.
+	pub fn do_validate_nested(
+		&self,
+		call: &<T as frame_system::Config>::RuntimeCall,
+		maybe_next_app_id: &mut Option<AppId>,
+	) -> Result<(), TransactionValidityError> {
+		let done = match call.is_sub_type() {
+			Some(UtilityCall::<T>::batch { calls })
+			| Some(UtilityCall::<T>::batch_all { calls })
+			| Some(UtilityCall::<T>::force_batch { calls }) => {
+				for call in calls.iter() {
+					let cast = call.clone().into();
```

Comment:
I am worry about this, because it duplicates `call` in memory in a recursive call, so we could need something like factorial of `max_extrinsic_len`. 
I mean, an attacker could create a nested call like:
```
// Number of nested `batchs` is `(max_extrinsic_len - remark_len) / batch_len`.
batch( batch( batch( ... remark)))
```
An it would consume the memory of the node.

I think there are some options here (like "The Good, The Bad, and the Ugly" :D ):
- a) Keep the recursive version, but avoid the clone in some way (references + lifetimes). Also c)
- b) Refactor it to an iterative solution. Also c)
- c) Just limit the number of recursions (or supported nested callls).

I think that implementing only option `c` is not good enough because of memory footprint could be really high even with a low number of max nested calls.

Reply:
@fmiguelgarcia I was also wondering, since this is for appIds and they will eventually be controlled by some entities. What would be the side effect of just allowing batches to have any valid appIds ? That would definitely simplify the issue

Reply:
I added a temporary solution where we do an actual cast. So far it's unsafe but I hope that @Aphoh could help me to make it safe.

I checked it with the debugger and everything seemed alright. Executing `p call` and `p c` on line 74 gave me the same output: `(&da_runtime::RuntimeCall) call = 0x00007fce580026b0`

Also I have added a limit to the maximum amount of recursion that can happen.

Reply:
Removed unsafe code. Utility::RuntimeCall implements trait `IsType<<Self as frame_system::Config>::RuntimeCall>` which means that is trivial to convert to and back from `< frame_system::Config>::RuntimeCall`